wire-ios-data-model/Source/Proteus/ProteusService.swift

```diff
@@ -173,31 +185,36 @@ public final class ProteusService: ProteusServiceInterface {
         if sessionExists(id: id) {
             logger.info("session exists, decrypting...")
 
-            do {
-                let decryptedBytes = try coreCrypto.perform { try $0.proteusDecrypt(
-                    sessionId: id.rawValue,
-                    ciphertext: data.bytes
-                )}
-
-                return (didCreateSession: false, decryptedData: decryptedBytes.data)
-            } catch {
-                logger.error("failed to decrypt data: \(String(describing: error))")
-                throw DecryptionError.failedToDecryptData
+            let decryptedBytes: Bytes = try coreCrypto.perform {
+                do {
+                    return try $0.proteusDecrypt(
+                        sessionId: id.rawValue,
+                        ciphertext: data.bytes
+                    )
+                } catch {
+                    logger.error("failed to decrypt data: \(error.localizedDescription)")
+                    throw DecryptionError.failedToDecryptData($0.lastProteusError)
+                }
             }
+
+            return (didCreateSession: false, decryptedData: decryptedBytes.data)
+
         } else {
             logger.info("session doesn't exist, creating one then decrypting message...")
 
-            do {
-                let decryptedBytes = try coreCrypto.perform { try $0.proteusSessionFromMessage(
-                    sessionId: id.rawValue,
-                    envelope: data.bytes
-                )}
-
-                return (didCreateSession: true, decryptedData: decryptedBytes.data)
-            } catch {
-                logger.error("failed to establish session from message: \(String(describing: error))")
-                throw DecryptionError.failedToEstablishSessionFromMessage
+            let decryptedBytes: Bytes = try coreCrypto.perform {
+                do {
+                    return try $0.proteusSessionFromMessage(
+                        sessionId: id.rawValue,
+                        envelope: data.bytes
+                    )
+                } catch {
+                    logger.error("failed to establish session from message: \(String(describing: error))")
+                    throw DecryptionError.failedToEstablishSessionFromMessage($0.lastProteusError)
+                }
```

Comment:
@typfel I've only mapped the error codes when decrypting a message. What do you do in android? Map error codes for all operations?

Reply:
We map it on all operation yes.

Reply:
You can see how every CC operation is wrapped in a `wrapException`:
 https://github.com/wireapp/kalium/blob/88503403b983ee2d5f3df4ecd4072f869f015018/cryptography/src/commonJvmAndroid/kotlin/com.wire.kalium.cryptography/ProteusClientCoreCryptoImpl.kt#L190